CUDA graphs overrides dynamic shapes and forces specialization

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103290

Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov @anijain2305

ezyang committed:
CUDA graphs overrides dynamic shapes and forces specialization

Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -307,14 +307,30 @@ def compile_fx_inner(
+        if not cudagraph_fail_reasons:
+            # Force specialize all inputs so that CUDA graphs will work
+            for t in example_inputs:
+                if isinstance(t, torch.SymInt):
```
do we need to do this for SymFloat as well ?

ezyang replied:
The contract is that SymFloat will never be an input per https://docs.google.com/document/d/1INSCdYu1PxXcr43HrD82OudeEuS-qxQe1yZmLg2wy6A/edit#heading=h.aytveihghlgn

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -307,14 +307,30 @@ def compile_fx_inner(
+                all(
+                    isinstance(t, (torch.Tensor, torch.SymInt)) for t in example_inputs
+                ),
+                "non-Tensor inputs",
```
We're forcing specialization for the sym ints... would dynamic shapes ever cause any other sorts of inputs to show up that wouldn't otherwise ? SymFloats ? ints ?

ezyang replied:
No, this is part of the contract.

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -1396,7 +1409,10 @@ def _allocate_and_copy_recording_inputs(self, inputs):
             for i, inp in enumerate(inputs):
-                if i not in self.static_input_idxs:
+                if not isinstance(inp, torch.Tensor):
+                    assert isinstance(inp, (int, torch.SymInt))
```
if we see an int here, that value won't be specialized right ? don't we want only symints ?

ezyang replied:
I suspect that the int case is impossible, but yeah, if an int does show up, that means we already specialized it and there's nothing to do

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -307,14 +307,30 @@ def compile_fx_inner(
+                    isinstance(t, (torch.Tensor, torch.SymInt)) for t in example_inputs
+                ),
+                "non-Tensor inputs",
+            ),
```
Also do we need to worry about non-Tensor outputs ?

ezyang replied:
Hmm, good question. SymInt outputs are permissible with dynamic shapes. Today, these would be static if all the inputs were static. So it seems in principle it should work.

ezyang replied:
I don't see how you can assert this? You would have to run the original graph to get what the true sizes are, but that would defeat the purpose of cudagraphs??

eellison reviewed (commented):
cool ! i have a few questions

ezyang commented:
@pytorchbot merge -f "spurious failures only"

eellison reviewed (commented):
You're going to need to update https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L583-L589 as well

eellison commented:
I think this was merged a bit prematurely.

vadimkantorov commented:
are recompilations cached? basically if we have a small collection of fixed shapes will the graphs/recompilations be cahed?

e.g. if I have a sequence: b1, b2, b1, b2 where b1 and b2 are input shapes, will it do compilation only twice?

ezyang commented:
@pytorchbot revert -c weird -m "to handle the other cuda graphs case"

ezyang commented:
> are recompilations cached? basically if we have a small collection of fixed shapes will the graphs/recompilations be cahed?

Yes, they're cached, up to the cache limit (which is 64 by default)

vadimkantorov commented:
> > are recompilations cached? basically if we have a small collection of fixed shapes will the graphs/recompilations be cahed?
> 
> Yes, they're cached, up to the cache limit (which is 64 by default)

i think it would be best to explain this directly in https://pytorch.org/docs/stable/generated/torch.compile.html?highlight=torch+compile#torch.compile

btw, the docs online do not have 2.0.1 version of pytorch! probably a bug!

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -332,6 +348,7 @@ def compile_fx_inner(
                 is_inference=is_inference,
             )
         else:
+            log.debug("disabled cudagraphs because %s", cudagraph_fail_reasons)
```
Wait wait I think you misunderstood. I just noticed there is preexisting code reporting cudagraphs disablement:

```
            if len(set(graph.device_types)) > 1:
                perf_hint_log.warning("skipping cudagraphs due to multiple devices")
            elif set(graph.device_types) == {"cuda"}:
                if graph.mutated_inputs:
                    perf_hint_log.warning("skipping cudagraphs due to input mutation")
                elif complex_memory_overlap_inputs:
                    perf_hint_log.warning(
                        "skipping cudagraphs due to complex input striding"
                    )
                elif len(graph.device_idxs) > 1 and config.triton.cudagraph_trees:
                    perf_hint_log.warning(
                        "skipping cudagraphs due to multiple device indexes"
                    )
```

This log message here is purely additive; I added it because I ran TORCH_LOGS=+inductor and I didn't see any log telling me if cudagraphs was running or not.  I did not modify the original `perf_hint_logs`.

I'm not really in the mood for bikeshedding, so tell me exactly what you want and I'll do it. If necessary I will just delete this debug line to unblock.

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented:
@eellison all fixed up and I added a test myself

ezyang commented:
@vadimkantorov I added more text to the doc, PTAL

vadimkantorov commented on torch/__init__.py:
```diff
@@ -1584,7 +1578,12 @@ def compile(model: Optional[Callable] = None, *,
-       dynamic (bool): Use dynamic shape tracing
+       dynamic (bool): Use dynamic shape tracing.  When this is True, we will up-front attempt
+        to generate a kernel that is as dynamic as possible to avoid recompilations when
+        sizes change.  This may not always work as some operations/optimizations will
```
hmm, maybe I miss where the docs were added, but this does not explain about cuda graphs / their requirements / compilation cache / how to change its size. how to track if cuda graphs were used and when recompilations happen

ezyang replied:
updated again, PTAL

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

vadimkantorov commented:
@ezyang thanks much for adding the explicative note :) now it's more clear. I think CUDA graphs are relatively obscure for now, so explaining more like this and e.g. mentioning their practicality for training / inference would make them be used by more people. btw why their switch is called quite unspecifically "reduce-overhead"? because from naming I'm inclined to always use reduce-overhead, i.e. why would i want more overhead :)

ezyang commented:
It's because you're trading off memory usage for overhead! So it's not a slam dunk.

vadimkantorov commented:
Yeah, I understand :) just that it's not clear for non-core-dec and a bit generic (can cuda graphs be useful outside of memory economizing mode? Or are they entangled? Is it about kernel launch time or memory overheads?)

And then the question becomes, should it be enabled by default?

ezyang commented:
Well, @eellison would certainly like that!

eellison reviewed (approved):
Thank you ! Looks great !

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -i

ezyang commented:
@pytorchbot merge -r main

ezyang committed:
Update on "CUDA graphs overrides dynamic shapes and forces specialization"


Previously, cudagraphs and dynamic_shapes were incompatible and enabling
dynamic shapes would forcibly disable cudagraphs.  This new strategy
I think is better.  The idea is essentially that cudagraphs is an
"optimization" that happens to guard on every input.  When cudagraphs
is on, we force everything static, and this automatically does the right
thing because we will force a recompile if sizes change.

This obsoletes https://github.com/pytorch/pytorch/pull/101813

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "i had a clean ci run from before with doc change only"
